pynestml/codegeneration/nest_builder.py

```diff
@@ -151,24 +148,35 @@ def build(self) -> None:
         else:
             shell = False
 
-        # first call cmake with all the arguments
-        try:
-            subprocess.check_call(cmake_cmd, stderr=stderr, stdout=stdout, shell=shell,
-                                  cwd=str(os.path.join(target_path)))
-        except subprocess.CalledProcessError as e:
-            raise GeneratedCodeBuildException(f"Error occurred during cmake! More detailed error messages can be found in {error_location}.")
-
-        # now execute make all
-        try:
-            subprocess.check_call(make_all_cmd, stderr=stderr, stdout=stdout, shell=shell,
-                                  cwd=str(os.path.join(target_path)))
-        except subprocess.CalledProcessError as e:
-            raise GeneratedCodeBuildException(f"Error occurred during 'make all'! More detailed error messages can be found in {error_location}.")
-
-        # finally execute make install
+        stages_exception = {"cmake": f"Error occurred during cmake! More detailed error messages can be found in {error_location}.", "build": f"Error occurred during 'make all'! More detailed error messages can be found in {error_location}.",
+                            "install":  f"Error occurred during 'make install'! More detailed error messages can be found in {error_location}."}
+        current_stage = ""
         try:
-            subprocess.check_call(make_install_cmd, stderr=stderr, stdout=stdout, shell=shell,
-                                  cwd=str(os.path.join(target_path)))
-        except subprocess.CalledProcessError as e:
-            raise GeneratedCodeBuildException(
-                f"Error occurred during 'make install'! More detailed error messages can be found in {error_location}.")
+            if self._options["redirect"]:
+                with open(self._options["stdout"], 'w') as stdout, open(self._options["stderr"], 'w') as stderr:
+                    current_stage = "cmake"
+                    subprocess.check_call(cmake_cmd, stderr=stderr, stdout=stdout, shell=shell,
+                                          cwd=str(os.path.join(target_path)))
+                    current_stage = "build"
+                    subprocess.check_call(make_all_cmd, stderr=stderr, stdout=stdout, shell=shell,
+                                          cwd=str(os.path.join(target_path)))
+                    current_stage = "install"
+                    subprocess.check_call(make_install_cmd, stderr=stderr, stdout=stdout, shell=shell,
+                                          cwd=str(os.path.join(target_path)))
+
+            else:
```

Comment:
Lot of code duplication between the if and the else parts. I would be happier with the old structure here. If no forwarding is requested, ``stdout`` and ``stderr`` could default to None, so the calls to ``subprocess.check_call()`` do not have to be repeated.

Reply:
I have removed the duplicated lines, and now we check if we should open the new files or not, and then execute the same block for both cases.

At the end, we close the files if they were opened already, either after finishing the CMake stages or after an exception.